File: Data Structures & Algorithms/max-area-of-island/submission-0.java
class Solution {
    private int[][] directions = new int[][]{{1,0}, {-1,0}, {0,1}, {0,-1}};
    private int m, n;

    private int dfs(int[][] grid, int row, int col){
        if(row < 0 || row == m || col < 0 || col == n || grid[row][col] == 0){
            return 0;
        }
        int sizeOfIsland = 1;
        grid[row][col] = 0;

        for(int[] direction : directions){
            int newRow = row + direction[0], newCol = col + direction[1];
            sizeOfIsland += dfs(grid, newRow, newCol);
        }
        return sizeOfIsland;
    }

    public int maxAreaOfIsland(int[][] grid) {
        int res = 0;
        m = grid.length;
        n = grid[0].length;

        for(int i = 0; i < m; ++i){
            for(int j = 0; j < n; ++j){
                if(grid[i][j] == 1){
                    res = Math.max(res, dfs(grid, i, j));
                }
            }
        }

        return res;
    }
}

File: Data Structures & Algorithms/rotting-fruit/submission-0.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        queue = deque()
        m, n = len(grid), len(grid[0])
        fresh, timeElapsed = 0, 0
        directions = [(0,1), (0,-1), (1,0), (-1,0)]

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    fresh += 1
                elif grid[i][j] == 2:
                    queue.append((i,j))

        while queue and fresh > 0:
            queueSize = len(queue)

            for _ in range(queueSize):
                row, col = queue.popleft()

                for direction in directions:
                    new_row, new_col = row + direction[0], col + direction[1]

                    if(0 <= new_row < m and 0 <= new_col < n and 
                        grid[new_row][new_col] == 1):
                        grid[new_row][new_col] = 2
                        fresh -= 1
                        queue.append((new_row, new_col))
            
            timeElapsed += 1
        
        return timeElapsed if fresh == 0 else -1